Stabilize doc_cfg, auto_doc_cfg and cfg_hide features

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/43781

For some reason I can't seem to reopen #79263 so opening a new one here.

The blocker was that `cfg(...)` didn't imply `doc(cfg(...))`, it was added through `doc_auto_cfg`. Discussion about it is below.

r? @rust-lang/rustdoc

GuillaumeGomez committed:
Stabilize doc_cfg, doc_auto_cfg and doc_cfg_hide features

Manishearth commented on src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:
```diff
@@ -276,3 +276,46 @@ To get around this limitation, we just add `#[doc(alias = "lib_name_do_something
 Users can now look for `lib_name_do_something` in our crate directly and find
 `Obj::do_something`.
+
+## `cfg`
```
kinda wish our docs here were better

Manishearth reviewed (approved):
r=me unless you're waiting for more people

Nemo157 commented:
I don't recall if there was an FCP previously?

`doc_auto_cfg` in particular I don't know that there was ever any discussion about making it active by default. We could potentially move that behind a `#![doc(auto_cfg)]` opt-in.

GuillaumeGomez commented on library/core/src/lib.rs:
```diff
@@ -173,7 +173,7 @@
 #![feature(decl_macro)]
 #![feature(deprecated_suggestion)]
-#![feature(doc_cfg)]
+#![cfg_attr(bootstrap, doc_cfg)]
```
This is what is creating the big mess. I'm finding out by what it was replaced (because we can't remove it altogether otherwise the first compilation stage will fail).

GuillaumeGomez commented:
> I don't recall if there was an FCP previously?

There wasn't.

> `doc_auto_cfg` in particular I don't know that there was ever any discussion about making it active by default. We could potentially move that behind a `#![doc(auto_cfg)]` opt-in.

It's an interesting point. Do you have an idea why someone wouldn't want `auto_cfg` to not be opt-in?

Nemo157 commented:
> It's an interesting point. Do you have an idea why someone wouldn't want `auto_cfg` to not be opt-in?

Some crates may not be setup to show this information yet, they may be full of internal `cfg` that they don't want users to see (e.g. used for compiler feature detection). If we made it opt-in (and maybe enabled by default in edition 2024) then they would have time to update to support it better.

> EDIT: I'd also like to precise that if `auto_cfg` is still under debate, I can opt it out of this stabilization until we figure it out.

I would prefer not to stabilize `doc(cfg)` without `auto_cfg`, that would result in a lot of churn adding the `doc(cfg)` attributes then removing them again when `auto_cfg` stabilizes. I think we should be able to quite quickly decide on whether anything about `auto_cfg` needs changing.

jyn514 commented:
> It seems that all the blockers were actually fixed a while ago so it seems like we can finally go forward with this.

Please say what the previous blockers were and how they have been fixed.

jyn514 commented:
This is the procedure for stabilizing a feature: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html

In particular, it would be very helpful to have a stabilization report: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#write-a-stabilization-report

GuillaumeGomez commented:
All valid points. Before starting the FCP, we need to discuss about whether or not we want `doc_auto_cfg` to be enabled by default or not and if we want an attribute to enable/disable it or not.

GuillaumeGomez commented:
Also I think we will have a problem with stdarch since they use the feature but I can't update it from this repository directly (as seen with the last CI error).

GuillaumeGomez commented:
The blocker was that `cfg(...)` didn't imply `doc(cfg(...))`, it was added through `doc_auto_cfg`. But since @Nemo157 mentioned the possibility to disable it (or not enable it by default), I looked at reasons at why we wouldn't want it and a very good example came up with: https://github.com/rust-lang/rust/issues/90497 (I'm also adding this into the first comment).

So I see a few possibilities from here:

 1. Enable the `doc_auto_cfg` by default and that's it. It has the advantage of not requiring a new argument, and since we also stabilize `cfg_hide`, people can handle it nicely. However, it requires extra actions from users without "warnings" about the change of behaviour.
 2. Add an attribute `#![doc(enable(auto_cfg))]` (or something along the line) which would enable the `doc_auto_cfg` feature. The problem with that is that it's very unlikely that many of our users will simply not know it exists and won't use it.
 3. Add an attribute `#![doc(disable(auto_cfg))]` (or something along the line) which would disable the `doc_auto_cfg` feature, meaning that the feature is enabled by default. The problem with having it enabled by default, like @Nemo157 said is that libraries will see a change in their output without "warnings".

Overall, my personal preference is `1.`: with `cfg_hide`, it's very easy to hide the information you don't want and pretty easy to do it. But `3.` also makes sense since the attribute would allow to simply disable the feature altogether.

So now: did I forget a case? If not, what is your preference?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7958854102?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/94c0675478dc996078203a2e5fb29fcfc0b93962/checks/7958854102/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling object v0.26.2
error: the feature `doc_cfg` has been stable since 1.65.0 and no longer requires an attribute to enable
  --> library/stdarch/crates/std_detect/src/lib.rs:16:33
   |
16 | #![feature(staged_api, stdsimd, doc_cfg, allow_internal_unstable)]
   |
   |
   = note: `-D stable-features` implied by `-D warnings`
error: could not compile `std_detect` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:04:16
```

</details>

euclio commented:
This feature still has some significant bugs: https://github.com/rust-lang/rust/issues/43781#issuecomment-939475077

GuillaumeGomez commented:
UI bugs except for https://github.com/rust-lang/rust/issues/83428. I think it'd be nice to fix them all before stabilizing though.

There is a suggested feature with https://github.com/rust-lang/rust/issues/73922 by adding this information into the search result too.

Overall, they don't impact the feature itself but the UI (except for https://github.com/rust-lang/rust/issues/83428 once again).

GuillaumeGomez commented:
As highlighted in https://github.com/rust-lang/rust/pull/101006, it seems like we need to have an attribute to enable/disable `doc_auto_cfg`. So two questions come from this:

 1. What should this attribute look like?
 2. Should it be enabled by default or not?

jplatte commented:
Re. 2, my 2c is it would be good to start off with off-by-default because it still requires manual work in some cases to look clean. Libraries that want to already put in a bit of extra work to document these things properly, switching the crate-level attribute will be easy enough.

Then with the 2024 Edition, the default can be reconsidered.

est31 commented:
:wave: Hello, I'm writing this comment in this **stabilization PR** to notify you, the authors of this PR, that https://github.com/rust-lang/rust/pull/100591 has been merged, which implemented a change in how features are stabilized.

Your PR has been filed before the change, so will likely require modifications in order to comply with the new rules. I recommend you to:

1. rebase the PR onto latest master, so that uses of the placeholder are possible.
2. replace the version numbers in the PR with the placeholder `CURRENT_RUSTC_VERSION`. For language changes, this means the version numbers in `accepted.rs` (example: 4caedbae869855a19ca49455e2cd338ae3b8c1ef). For library changes, this means the `since` fields (example e576a9b554f24a26c36c04b41c901636d190605b).

That's it! The `CURRENT_RUSTC_VERSION` placeholder will, as part of the release process, be replaced with the version number that the PR merged for. It can be used anywhere in `rust-lang/rust`, not just accepted.rs and the since fields.

If you have any questions, feel free to drop by the [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/libs.20stabilization.20placeholder), or ping me directly in this PR's thread. Thanks! :wave:

GuillaumeGomez commented:
@est31 Thanks for the heads-up! As jyn514 mentioned it, I also need to write a small recap' of the PR and I think we need to go through an FCP too. But in any case, we need to agree on the attribute syntax to add for `doc_auto_cfg`.

GuillaumeGomez committed:
Add doc(auto_cfg) and doc(no_auto_cfg) attributes

GuillaumeGomez committed:
Add missing `#![doc(auto_cfg)]` attribute in rustdoc tests

GuillaumeGomez committed:
Add rustdoc-ui test for `doc(auto_cfg)` and `doc(no_auto_cfg)` attributes

GuillaumeGomez committed:
Add documentation for `doc(auto_cfg)` and `doc(no_auto_cfg)` attributes

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3129331632/jobs/5078288786) [(plain)](https://github.com/rust-lang/rust/commit/092d4ed08786bf9f53347d72ac233b61a69c4ccc/checks/8554796328/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling hashbrown v0.12.3
error: the feature `doc_cfg` has been stable since 1.65.0 and no longer requires an attribute to enable
  --> library/stdarch/crates/std_detect/src/lib.rs:16:33
   |
16 | #![feature(staged_api, stdsimd, doc_cfg, allow_internal_unstable)]
   |
   |
   = note: `-D stable-features` implied by `-D warnings`
error: could not compile `std_detect` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:03:50
```

</details>

GuillaumeGomez commented:
I'll need to send a PR to `stdarch` before to update it before we can merge this PR. But can be done after the FCP.

So I updated this PR with the following changes:
 * Added the `doc(auto_cfg)` and `doc(no_auto_cfg)` attributes.
 * I added the automatic switch for the 2024 edition when `auto_cfg` will become the default behaviour.
 * I added tests for the attributes.

Does it look ok to everyone or is there something else which needs to be done before opening the FCP?

jhpratt commented:
Sounds reasonable to me.

GuillaumeGomez commented on src/librustdoc/visit_ast.rs:
```diff
@@ -146,6 +147,22 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
+            .iter()
+            .filter(|attr| attr.has_name(sym::doc))
+            .flat_map(|attr| attr.meta_item_list().into_iter().flatten())
+            .find(|attr| attr.has_name(sym::auto_cfg) || attr.has_name(sym::auto_cfg))
```
Great catch!

GuillaumeGomez commented on src/test/rustdoc-ui/doc_no_auto_cfg.rs:
```diff
@@ -0,0 +1,13 @@
+/*#![doc(no_auto_cfg(1))] //~ ERROR
+#![doc(no_auto_cfg = 1)] //~ ERROR
+#![doc(auto_cfg(1))] //~ ERROR
+#![doc(auto_cfg = 1)] //~ ERROR
```
I was hesitant about adding this check but I guess it doesn't hurt. :3

GuillaumeGomez committed:
Add check for usage of both `doc(auto_cfg)` and `doc(no_auto_cfg)` at the same time

GuillaumeGomez committed:
Add tests for duplicated usage of `doc(auto_cfg)` and `doc(no_auto_cfg)`

GuillaumeGomez commented:
I added the extra checks (usage of both `auto_cfg` and `no_auto_cfg` at the same time) suggested by @notriddle. If it's ok for everyone, I'll write the recap and start an FCP in the next days.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3137666021/jobs/5096100575) [(plain)](https://github.com/rust-lang/rust/commit/2d06d6506652c3d0281fccf291c54d22bfe37766/checks/8578635117/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
error: the feature `doc_cfg` has been stable since 1.65.0 and no longer requires an attribute to enable
  --> library/stdarch/crates/std_detect/src/lib.rs:16:33
   |
16 | #![feature(staged_api, stdsimd, doc_cfg, allow_internal_unstable)]
   |
   |
   = note: `-D stable-features` implied by `-D warnings`
error: could not compile `std_detect` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:04:19
```

</details>

GuillaumeGomez commented:
## Stabilization Report

### Implementation History

- Initial implementation in https://github.com/rust-lang/rust/pull/43348.
- `doc_cfg_hide` added in https://github.com/rust-lang/rust/pull/89596.
- `doc_auto_cfg` added in https://github.com/rust-lang/rust/pull/90502.
- Unified `doc(cfg())` and `cfg()` syntax in https://github.com/rust-lang/rust/pull/84442.
- Adjusted documentation in https://github.com/rust-lang/rust/pull/92818.
- Improved rendering of crate features in https://github.com/rust-lang/rust/pull/75330.
- Simplified rendering based on current context in https://github.com/rust-lang/rust/pull/77672.
- Improved rendering on traits in https://github.com/rust-lang/rust/pull/78678.
- Removed `cfg(doc)` and `cfg(doctest)` from `doc_auto_cfg` and fixed `not()` making badge disappear in https://github.com/rust-lang/rust/pull/98108.
- Added the doc(auto_cfg) and doc(no_auto_cfg) attributes (in this PR).
- Added the automatic switch for the 2024 edition when auto_cfg will become the default behaviour (in this PR).

### Summary

These attributes allow to give extra information to the documentation readers about the conditions (platform, features, etc) under which an item is available. Basically, the goal is to show to users the `cfg()` used on items directly into the documentation. Starting the 2024 edition, this information will be retrieved automatically unless `doc(no_auto_cfg)` is used.

### Experience Report

With this feature, crates owners will have less to do in order to provide the availability conditions of a given item while documentation readers will get more information.

GuillaumeGomez commented:
Let's start the FCP then. I already talked with @Amanieu about how we will go around the incompatibility with `stdarch` and he told me that we can force push a PR there which removes the feature so that it can fix the CI here.

@rfcbot merge

jyn514 commented:
> doc(no_auto_cfg)

What are the semantics of this attribute? Will it be crate level, or per-item? Is it possible to use in 2021 edition, or only 2024? If the former, can we implement it and add tests today?

> doc(auto_cfg)

What will happen to this in the 2024 edition? Will it become a no-op? If so, will rustdoc lint when it's present? What will happen when both this and `no_auto_cfg` are present?

Where do the tests for all these features live, including the attributes added in this PR?

As a general comment, I think it is a bad idea to add new features at the same time a feature is stabilized (for anything, not just rustdoc). If we think new attributes are necessary for a good UX, we should add those, let them bake on nightly for a cycle, and only then stabilize the feature. We did this for `--show-doctest-warnings` and found that the feature could be replicated using only existing stable features.

GuillaumeGomez commented:
> > doc(no_auto_cfg)
> 
> What are the semantics of this attribute? Will it be crate level, or per-item? Is it possible to use in 2021 edition, or only 2024? If the former, can we implement it and add tests today?

Both attributes exist and are tested. The edition switch is also implemented. Tests and implementation are in the last 4 commits.

> > doc(auto_cfg)
> 
> What will happen to this in the 2024 edition? Will it become a no-op? If so, will rustdoc lint when it's present? What will happen when both this and `no_auto_cfg` are present?

It emis a warning if it's used whereas it's the default so `doc(no_auto_cfg)` emits a warning before the 2024 edition and the opposite for `doc(auto_cfg)`.

When both are present, it emits an error (tested as well).

> Where do the tests for all these features live, including the attributes added in this PR? Where does the documentation live?

You can see them added in the last 4 commits. Documentation is in the rustdoc book in the `doc` attributes chapter. Tests are both in `rustdoc-ui` (for rustdoc output) and in `rustdoc` for usage of `auto_cfg`.

> As a general comment, I think it is a bad idea to add new features at the same time a feature is stabilized (for anything, not just rustdoc). If we think new attributes are necessary for a good UX, we should add those, let them bake on nightly for a cycle, and only then stabilize the feature. We did this for `--show-doctest-warnings` and found that the feature could be replicated using only existing stable features.

I agree but in this case, it's a bit different because currently, `auto_cfg` is enabled through the feature. It would "break" the crates using this feature if we only add the attribute becaude they would need to add the attribute in addition to the feature to get the same result and it would "fail" silently, which is quite bad. By stabilizing the feature alongside the two new attributes, there is no break, just need to replace how the feature is enabled (which is expectef when a feature is stabilized). Does it make sense to you?

Nemo157 commented:
> there is no break, just need to replace how the feature is enabled 

This is exactly the same level of breakage for crates using it. Either way they have to add the new attribute, the only difference is whether they can simultaneously remove the `feature`.

GuillaumeGomez commented:
No. Adding the `auto_cfg` will silently disable the feature for the crates using it. It's not the same level. However in term of attributes updates, it's more or less the same.

Just to be clear: I'm fine with doing it in another PR, but wouldn't we need to add a warning to tell users that the default behavour changed?

Nemo157 commented:
> wouldn't we need to add a warning to tell users that the default behavour changed?

Yeah, I also just had this idea and edited it into my comment as you wrote yours 😅 We wouldn't _need_ to since this is unstable still, but it would be nice to since it should be an easy to add warning (but we can only have that warning while the feature is unstable, we can't add it if we also stabilize it at the same time).

GuillaumeGomez commented:
Sounds good to me. I'll open a PR adding these two attributes there and make a warning. Next release cycle, we can go forward on this. I'll leave the FCP open in the meantime.

Manishearth commented:
I think in that case we should cancel the FCP.

jmillikin commented:
It sounds like the lack of consensus is around `doc(auto_cfg)` and `doc(no_auto_cfg)` -- would it be possible to leave those unstabilized for now, and only stabilize `doc(cfg(...))` ?

My understanding is that the `doc_cfg` feature doesn't have any backwards-compatibility concerns, no new warnings, and no interaction with the editions system. It's purely additive, and gives library authors an important capability that is currently missing in stable.

Nemo157 commented:
I mentioned earlier:

> I would prefer not to stabilize `doc(cfg)` without `auto_cfg`, that would result in a lot of churn adding the `doc(cfg)` attributes then removing them again when `auto_cfg` stabilizes.

From what I have seen with both features available usage of `doc(cfg)` is rare, the auto detected cfg is normally all that is needed.

jmillikin commented:
As one data point, I have no plans to use `auto_cfg` within my own code regardless of its stability.

If there is a set of developers who would prefer to avoid code churn, then they have the option of not using `doc(cfg(..))`. Allowing it to stabilize won't make them worse off (they'll be waiting for `auto_cfg` either way), and it is beneficial to folks who are OK with the non-auto functionality.

GuillaumeGomez commented:
`auto_cfg` doesn't remove the need for `doc(cfg)`. In some cases, you don't want `auto_cfg` at all and manuall do `doc(cfg)` whereas in some other cases, you just want to add some extra information that is not provided automatically with `auto_cfg`. I think both are needed. But yes, I can cancel the FCP if it's better to wait for the new release to be out.

@rfcbot cancel

jhpratt commented:
For what it's worth, while I previously used `doc_cfg` directly (alongside `doc_auto_cfg`), earlier tonight I checked every instance where I used `doc_cfg` explicitly. Turns out `doc_auto_cfg` had been improved sufficiently such that `doc_cfg` was not necessary for my use case (the `time` crate). While I can imagine that there's still _a_ use case, it's likely vastly smaller than it used to be because of the improvements I noticed.

(also, thank you to those that made the improvements)

GuillaumeGomez commented:
Glad to hear it! As for an example, the `std::fs` module is having issues with `doc_auto_cfg` iirc because the mix between `cfg(doc)` and `auto_cfg` gives pretty bad results.

GuillaumeGomez commented:
So what should we do about this and about https://github.com/rust-lang/rust/pull/102599 ? I think stabling `doc(cfg)` without `auto_cfg` wouldn't be the best approach, however, having a way to enable/disable `auto_cfg` seems like a requirement too.

jhpratt commented:
Maybe land the attributes and wait one release cycle to make sure there's no concerns? Then it can all be stabilized together.

GuillaumeGomez commented:
It's already the plan. The attributes are in https://github.com/rust-lang/rust/pull/102599. I was mostly asking if it looked like what they expected and if they thought if it was ready to be merged. ;)

jhpratt commented:
In my opinion, yes.

GuillaumeGomez commented:
We need someone from the @rust-lang/rustdoc team to agree with us then.

GuillaumeGomez commented:
Sure, let me explain the situation a bit more then. The stabilization report written [here](https://github.com/rust-lang/rust/pull/100883#issuecomment-1264403140) is still valid, only the way we're moving forward changed. Once #102599 merged, I planned to write another one before re-opening the FCP. However you're right, more context will make everyone's life easier:

In this PR, we want to merge the `doc_cfg`, `doc_auto_cfg` and `cfg_hide` features (as they're all co-dependant). The problem was that stabilizing `doc_auto_cfg` meant that the `cfg`s were retrieved automatically, which was an issue for some people. Someone suggested that we added attributes to enable/disable this feature. The discussion about these two new attributes happened [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Add.20new.20attribute.20to.20enable.2Fdisable.20doc_auto_cfg).

The conclusions were as follow: until the 2024 edition, `doc(no_auto_cfg)` will be enabled by default so if you want the feature, you'll need to use `doc(auto_cfg)`. If your code is under the 2024 edition though, it's the opposite: `doc(auto_cfg)` is enabled by default.

Now the problem was that we needed to let users be aware that using `feature(doc_auto_cfg)` was not enough anymore to get this feature working, they also need `doc(auto_cfg)`. So we will have one release cycle where they will have a warning before we merge this PR, stabilizing these features. This is what #102599 is for.

Manishearth commented:
> In this PR, we want to merge the `doc_cfg`, `doc_auto_cfg` and `cfg_hide` features (as they're all co-dependant).

Can you please illustrate, _in detail_, what each of these features does? I think that's a major part of the confusion here, we all have a general idea of them but not a precise idea of what separates them and their status quo.

And again, ideally this is all one single doc so people do not need to trawl through multiple discussions to review it. It might take a while to write, that's fine.

GuillaumeGomez commented:
Thanks for the suggestion! I opened a zulip thread [here](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Discussion.20about.20.60doc_cfg*.60.20features.20stabilization).

bors commented:
:umbrella: The latest upstream changes (presumably #102948) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

3tilley commented:
For anyone else following from afar, this is the best description I've found of the [issues](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Discussion.20about.20.60doc_cfg*.60.20features.20stabilization/near/304190110).

jyn514 commented:
There are a lot of conflicts here. I'm going to close the PR for now by the same rationale as https://github.com/rust-lang/rust/pull/102599#issuecomment-1522868483, since this is blocked on an RFC.
